File: tools/HME/proxy/project_adapter.js
'use strict';

const fs = require('fs');
const path = require('path');
const { PROJECT_ROOT } = require('./hme_paths');

const DEFAULT_ADAPTER = Object.freeze({
  project_id: 'generic',
  project_name: 'Generic project',
  domain: 'software',
  source_roots: ['src'],
  project_docs: ['doc/composition.md'],
  primary_doc: 'doc/composition.md',
  pipeline: { main: 'npm test' },
  artifacts: { metrics_dir: 'src/output/metrics' },
  optional_artifacts: [],
  capabilities: {},
  health: {},
});

function adapterPath(root = PROJECT_ROOT) {
  return process.env.HME_PROJECT_ADAPTER || path.join(root, 'config', 'project-adapter.json');
}

function readJson(file) {
  try { return JSON.parse(fs.readFileSync(file, 'utf8')); }
  catch (_err) { return null; }
}

function mergeAdapter(raw) {
  const cfg = { ...DEFAULT_ADAPTER, ...(raw || {}) };
  cfg.source_roots = Array.isArray(cfg.source_roots) ? cfg.source_roots : ['src'];
  cfg.project_docs = Array.isArray(cfg.project_docs) ? cfg.project_docs : [cfg.primary_doc];
  cfg.pipeline = cfg.pipeline && typeof cfg.pipeline === 'object' ? cfg.pipeline : {};
  cfg.artifacts = cfg.artifacts && typeof cfg.artifacts === 'object' ? cfg.artifacts : {};
  cfg.optional_artifacts = Array.isArray(cfg.optional_artifacts) ? cfg.optional_artifacts : [];
  cfg.capabilities = cfg.capabilities && typeof cfg.capabilities === 'object' ? cfg.capabilities : {};
  cfg.health = cfg.health && typeof cfg.health === 'object' ? cfg.health : {};
  return cfg;
}

function loadAdapter(root = PROJECT_ROOT) {
  return mergeAdapter(readJson(adapterPath(root)));
}

function resolvePath(root, relPath) {
  const rel = String(relPath || '');
  const abs = path.resolve(root, rel);
  const back = path.relative(path.resolve(root), abs);
  if (back.startsWith('..') || path.isAbsolute(back)) {
    throw new Error(`adapter path escapes project root: ${rel}`);
  }
  return abs;
}

function artifactPath(name, root = PROJECT_ROOT, adapter = loadAdapter(root)) {
  const rel = adapter.artifacts && adapter.artifacts[name];
  return rel ? resolvePath(root, rel) : '';
}

function sourceRoots(root = PROJECT_ROOT, adapter = loadAdapter(root)) {
  return adapter.source_roots.map((p) => resolvePath(root, p));
}

function projectDocs(root = PROJECT_ROOT, adapter = loadAdapter(root)) {
  return adapter.project_docs.map((p) => resolvePath(root, p));
}

function hasCapability(name, adapter) {
  return Boolean((adapter || loadAdapter()).capabilities[name]);
}

module.exports = {
  DEFAULT_ADAPTER,
  adapterPath,
  loadAdapter,
  resolvePath,
  artifactPath,
  sourceRoots,
  projectDocs,
  hasCapability,
};

File: tools/HME/scripts/project_adapter.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path

from hme_paths import PROJECT_ROOT


@dataclass
class ProjectAdapter:
    project_id: str = ""
    project_name: str = ""
    domain: str = ""
    source_roots: list = field(default_factory=list)
    project_docs: list = field(default_factory=list)
    primary_doc: str = ""
    pipeline: dict = field(default_factory=dict)
    artifacts: dict = field(default_factory=dict)
    optional_artifacts: list = field(default_factory=list)
    capabilities: dict = field(default_factory=dict)
    health: dict = field(default_factory=dict)


DEFAULT_ADAPTER = {
    "project_id": "generic",
    "project_name": "Generic project",
    "domain": "software",
    "source_roots": ["src"],
    "project_docs": ["doc/composition.md"],
    "primary_doc": "doc/composition.md",
    "pipeline": {"main": "npm test"},
    "artifacts": {"metrics_dir": "src/output/metrics"},
    "optional_artifacts": [],
    "capabilities": {},
    "health": {},
}


def adapter_path(root: Path = PROJECT_ROOT) -> Path:
    override = os.environ.get("HME_PROJECT_ADAPTER")
    return Path(override).expanduser() if override else root / "config" / "project-adapter.json"


def load_adapter(root: Path = PROJECT_ROOT) -> dict:
    path = adapter_path(root)
    raw = {}
    if path.is_file():
        raw = json.loads(path.read_text(encoding="utf-8"))
    cfg = {**DEFAULT_ADAPTER, **raw}
    cfg["source_roots"] = cfg.get("source_roots") or ["src"]
    cfg["project_docs"] = cfg.get("project_docs") or [cfg.get("primary_doc", "doc/composition.md")]
    cfg["pipeline"] = cfg.get("pipeline") or {}
    cfg["artifacts"] = cfg.get("artifacts") or {}
    cfg["optional_artifacts"] = cfg.get("optional_artifacts") or []
    cfg["capabilities"] = cfg.get("capabilities") or {}
    cfg["health"] = cfg.get("health") or {}
    return cfg


def resolve_path(root: Path, rel_path: str) -> Path:
    candidate = (root / rel_path).resolve()
    candidate.relative_to(root.resolve())
    return candidate


def artifact_path(name: str, root: Path = PROJECT_ROOT, adapter: dict | None = None) -> Path | None:
    cfg = adapter or load_adapter(root)
    rel = cfg.get("artifacts", {}).get(name)
    return resolve_path(root, rel) if rel else None


def source_roots(root: Path = PROJECT_ROOT, adapter: dict | None = None) -> list[Path]:
    cfg = adapter or load_adapter(root)
    return [resolve_path(root, p) for p in cfg.get("source_roots", [])]


def project_docs(root: Path = PROJECT_ROOT, adapter: dict | None = None) -> list[Path]:
    cfg = adapter or load_adapter(root)
    return [resolve_path(root, p) for p in cfg.get("project_docs", [])]


def has_capability(name: str, root: Path = PROJECT_ROOT, adapter: dict | None = None) -> bool:
    cfg = adapter or load_adapter(root)
    return bool(cfg.get("capabilities", {}).get(name))
